[AOTInductor] Bug fix for freeing buffers when freeing multiple times

muchulee8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149810

Summary:
We might free the active buffer if we free the buffer twice.

Test Plan:
```
LD_LIBRARY_PATH=/data/users/$USER/pytorch/build/lib
/home/$USER/local/pytorch/build/bin/test_aoti_inference
```
Reviewers:

Subscribers:

Tasks:

Tags:

muchulee8 committed:
[AOTInductor] Bug fix for freeing buffers when freeing multiple times

Summary:
We might free the active buffer if we free the buffer twice.

Test Plan:
LD_LIBRARY_PATH=/data/users/$USER/pytorch/build/lib
/home/$USER/local/pytorch/build/bin/test_aoti_inference

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

chenyang78 commented on test/cpp/aoti_inference/test.cpp:
```diff
@@ -405,12 +405,19 @@ void test_aoti_free_buffer() {
+  }
+  ASSERT_EQ(initMemory, updateMemory1);
+
+  // Free buffer #1 again, since #1 is freed, nothing should change.
```
Out of curiosity, is freeing an already-freed buffer well defined behavior here?

muchulee8 replied:
The buffer is actually just a unique_ptr, so freeing it again should just be a no-op. So the change in model_container we just "pass-through" what users are doing, resetting the unique_ptr.

chenyang78 replied:
Hmm, could you double-check? I think double-freeing a unique_ptr is undefined behavior in C++.

chenyang78 replied:
Ah, I see. So we pass a nullptr to the reset for the second call. Thanks.

muchulee8 committed:
Update on "[AOTInductor] Bug fix for freeing buffers when freeing multiple times"


Summary:
We might free the active buffer if we free the buffer twice.

Test Plan:
```
LD_LIBRARY_PATH=/data/users/$USER/pytorch/build/lib
/home/$USER/local/pytorch/build/bin/test_aoti_inference
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

muchulee8 commented:
@pytorchbot merge
